apollo-runtime/src/main/java/com/apollographql/android/impl/CacheCallInterceptor.java

```diff
@@ -0,0 +1,176 @@
+package com.apollographql.android.impl;
+
+import com.apollographql.android.CustomTypeAdapter;
+import com.apollographql.android.api.graphql.Operation;
+import com.apollographql.android.api.graphql.Response;
+import com.apollographql.android.api.graphql.ResponseFieldMapper;
+import com.apollographql.android.api.graphql.ScalarType;
+import com.apollographql.android.cache.normalized.Cache;
+import com.apollographql.android.cache.normalized.CacheControl;
+import com.apollographql.android.cache.normalized.CacheKeyResolver;
+import com.apollographql.android.cache.normalized.ReadableCache;
+import com.apollographql.android.cache.normalized.Record;
+import com.apollographql.android.cache.normalized.ResponseNormalizer;
+import com.apollographql.android.cache.normalized.Transaction;
+import com.apollographql.android.cache.normalized.WriteableCache;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+
+import javax.annotation.Nullable;
+
+final class CacheCallInterceptor implements CallInterceptor {
+  private final Cache cache;
+  private final CacheControl cacheControl;
+  private final ResponseFieldMapper responseFieldMapper;
+  private final Map<ScalarType, CustomTypeAdapter> customTypeAdapters;
+  private final ExecutorService dispatcher;
+
+  CacheCallInterceptor(Cache cache, CacheControl cacheControl, ResponseFieldMapper responseFieldMapper,
+      Map<ScalarType, CustomTypeAdapter> customTypeAdapters, ExecutorService dispatcher) {
+    this.cache = cache;
+    this.cacheControl = cacheControl;
+    this.responseFieldMapper = responseFieldMapper;
+    this.customTypeAdapters = customTypeAdapters;
+    this.dispatcher = dispatcher;
+  }
+
+  @Override public InterceptorResponse intercept(Operation operation, CallInterceptorChain chain) throws IOException {
+    InterceptorResponse cachedResponse = resolveCacheFirstResponse(operation);
+    if (cachedResponse != null) {
+      return cachedResponse;
+    }
+
+    InterceptorResponse networkResponse = null;
+    try {
+      networkResponse = chain.proceed();
+      return handleNetworkResponse(operation, networkResponse);
+    } catch (Exception e) {
+      InterceptorResponse response = resolveNetworkFirstCacheResponse(operation, networkResponse);
+      if (response != null) {
+        return response;
+      }
+      throw e;
+    }
+  }
+
+  @Override
+  public void interceptAsync(final Operation operation, final CallInterceptorChain chain,
```

Comment:
Makes sense. I always wondered why our internal code base had the dispatcher copied out, now I know why ☺️